src/middlewared/middlewared/plugins/snmp.py

```diff
@@ -1,10 +1,126 @@
 import middlewared.sqlalchemy as sa
+import subprocess
+import os
 
+from contextlib import suppress
+from cryptography.fernet import Fernet
 from middlewared.common.ports import ServicePortDelegate
 from middlewared.schema import Bool, Dict, Int, Password, Str
-from middlewared.service import SystemServiceService, ValidationErrors
+from middlewared.service import private, SystemServiceService, ValidationErrors, CallError
+from middlewared.utils.crypto import generate_string
 from middlewared.validators import Email, Match, Or, Range
 
+SNMP_PRIV_CONF = '/var/lib/snmp/snmpd.conf'
+SNMP_PRIV_KEY = '/var/lib/private/snmp'
+_SNMP_SYSTEM_USER = {
+    'name': 'snmpSystemUser', 'auth_type': 'SHA', 'key': None, 'size': 0
+}
+
+
+def _get_authuser_secret():
+    """
+    Get the auth user saved secret.
+    Internal helper function for use by this module.
+    Return decoded string.
+    """
+    secret = ""
+    if not _SNMP_SYSTEM_USER['key']:
+        # No system user key registered
+        return secret
+
+    try:
+        with open(SNMP_PRIV_KEY, 'rb') as fd:
+            secret_c = fd.read(_SNMP_SYSTEM_USER['size'])
+    except FileNotFoundError:
+        return secret
+
+    try:
+        secret = Fernet(_SNMP_SYSTEM_USER['key']).decrypt(secret_c).decode()
+    except TypeError:
+        return secret
+
+    return secret
+
+
+def _set_authuser_secret(secret):
+    """
+    Save the auth user secret.
+    Internal helper function for use by this module.
+    INPUT: ascii string (not encoded)
+    """
+    secret_c = Fernet(_SNMP_SYSTEM_USER['key']).encrypt(secret.encode())
+    with open(SNMP_PRIV_KEY, 'wb') as fd:
```

Comment:
Do we need to set permissions on this?

Reply:
os.fchmod(fd.fileno(), mode)

Reply:
I think it would be better to have this in a python script in `etc_files` and call into it via `etc.generate`. That way we can use standardized methods of generating config files.